### PR TITLE
Don't call `query_normalize` when reporting similar impls

### PR DESCRIPTION
**Pull request by compiler-errors**
Firstly, It's sketchy to be using `query_normalize` at all during HIR typeck -- it's asking for an ICE 😅. Secondly, we're normalizing an impl trait ref that potentially has parameter types in `ty::ParamEnv::empty()`, which is kinda sketchy as well.

The only UI test change from removing this normalization is that we don't evaluate anonymous constants in impls, which end up giving us really ugly suggestions:

```
error[E0277]: the trait bound `[X; 35]: Default` is not satisfied
 --> /home/gh-compiler-errors/test.rs:4:5
  |
4 |     <[X; 35] as Default>::default();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `[X; 35]`
  |
  = help: the following other types implement trait `Default`:
            &[T]
            &mut [T]
            [T; 32]
            [T; core::::array::{impl#30}::{constant#0}]
            [T; core::::array::{impl#31}::{constant#0}]
            [T; core::::array::{impl#32}::{constant#0}]
            [T; core::::array::{impl#33}::{constant#0}]
            [T; core::::array::{impl#34}::{constant#0}]
          and 27 others
```

So just fold the impls with a `BottomUpFolder` that calls `ty::Const::eval`. This doesn't work totally correctly with generic-const-exprs, but it's fine for stable code, and this is error reporting after all.

### CODE REVIEW

File: tests/ui/consts/missing-larger-array-impl.stderr [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+   |
+   = help: the following other types implement trait `Default`:
+             &[T]
+             &mut [T]
+             [T; 0]
+             [T; 10]
+             [T; 11]
+             [T; 12]
+             [T; 13]
+             [T; 14]
```
Comment (cjgillot) [inline tests/ui/consts/missing-larger-array-impl.stderr:15]: The order is quite funny here. Any idea why those are not in increasing order?

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -2028,27 +2027,26 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
-                    .at(&ObligationCause::dummy(), ty::ParamEnv::empty())
-                    .query_normalize(trait_ref)
-                    .map_or(trait_ref, |normalized| normalized.value);
-                (similarity, normalized)
-            })
-            .collect::<Vec<_>>();
-        normalized_impl_candidates_and_similarities.sort();
-        normalized_impl_candidates_and_similarities.dedup();
+        let mut impl_candidates = impl_candidates.to_vec();
+        impl_candidates.sort_by_key(|cand| (cand.similarity, cand.trait_ref));
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2031]: Is this useful? `report` sorts everything again.

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2031]: Removed the sort in `report` -- that should fix the weird rendering and also actually make this sort meaningful. It does have a lot of fallout, but the changes are either neutral or slightly better suggestions.

File: tests/ui/consts/missing-larger-array-impl.stderr (hunk 1 above)
Comment (compiler-errors) [reply tests/ui/consts/missing-larger-array-impl.stderr]: Because the code was sorting by string 😓

**Comment by cjgillot** [comment]
Great!
@bors r+

**Pull request merged**